pandas/core/frame.py

```diff
@@ -217,11 +217,16 @@ def __init__(self, data=None, index=None, columns=None, dtype=None,
         if isinstance(data, DataFrame):
             data = data._data
 
+        if hasattr(data, 'to_dataframe'):  # xr.Dataset
```

Comment:
The broader question is whether we want objects to be able to define their own conversion to a `DataFrame` - I think the main options are:
1. Everything is treated as a `dict`
2. We build the code within the `DataFrame` constructor - i.e. that `if` statement checks for `DataSet` (and anything else we want to check)
3. There is some duck-like method such as `to_dataframe` that classes can define themselves

Thoughts?